src/sabre_swap/sabre_dag.rs

```diff
@@ -34,36 +33,35 @@ impl SabreDAG {
         num_qubits: usize,
         num_clbits: usize,
         nodes: Vec<(usize, Vec<usize>, HashSet<usize>)>,
-        front_layer: PyReadonlyArray1<usize>,
     ) -> PyResult<Self> {
-        let mut qubit_pos: Vec<usize> = vec![usize::MAX; num_qubits];
-        let mut clbit_pos: Vec<usize> = vec![usize::MAX; num_clbits];
-        let mut reverse_index_map: HashMap<usize, NodeIndex> = HashMap::with_capacity(nodes.len());
+        let mut qubit_pos: Vec<Option<NodeIndex>> = vec![None; num_qubits];
+        let mut clbit_pos: Vec<Option<NodeIndex>> = vec![None; num_clbits];
         let mut dag: DiGraph<(usize, Vec<usize>), ()> =
             Graph::with_capacity(nodes.len(), 2 * nodes.len());
+        let mut first_layer = Vec::<NodeIndex>::new();
         for node in &nodes {
             let qargs = &node.1;
             let cargs = &node.2;
             let gate_index = dag.add_node((node.0, qargs.clone()));
-            reverse_index_map.insert(node.0, gate_index);
+            let mut is_front = true;
             for x in qargs {
-                if qubit_pos[*x] != usize::MAX {
-                    dag.add_edge(NodeIndex::new(qubit_pos[*x]), gate_index, ());
+                if let Some(predecessor) = qubit_pos[*x] {
+                    is_front = false;
+                    dag.add_edge(predecessor, gate_index, ());
                 }
-                qubit_pos[*x] = gate_index.index();
+                qubit_pos[*x] = Some(gate_index);
             }
             for x in cargs {
-                if clbit_pos[*x] != usize::MAX {
-                    dag.add_edge(NodeIndex::new(clbit_pos[*x]), gate_index, ());
+                if let Some(predecessor) = clbit_pos[*x] {
+                    is_front = false;
+                    dag.add_edge(predecessor, gate_index, ());
                 }
-                clbit_pos[*x] = gate_index.index();
+                clbit_pos[*x] = Some(gate_index);
+            }
+            if is_front {
+                first_layer.push(gate_index);
```

Comment:
I'm glad we're able to do this inline now. In #8388 I had originally something similar to generate the first layer as we iterated and ending up producing a slightly different but still valid result in some cases so I reverted to just using the `DAGCircuit` method for that. This will make further usage of this data structure from inside rust (for other callers to sabre swap) easier.